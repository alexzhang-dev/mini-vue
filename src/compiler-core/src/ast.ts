export const enum NodeType {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
}
