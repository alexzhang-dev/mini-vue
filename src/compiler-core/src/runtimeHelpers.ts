export const TO_DISPLAY_STRING = Symbol('toDisplayString')

export const HelperNameMapping = {
  [TO_DISPLAY_STRING]: 'toDisplayString',
}
