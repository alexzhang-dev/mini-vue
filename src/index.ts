export * from './runtime-core/index'
export * from './reactivity/index'
