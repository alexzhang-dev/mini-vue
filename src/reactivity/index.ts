export const sum = function (a: number, b: number) {
  return a + b
}
