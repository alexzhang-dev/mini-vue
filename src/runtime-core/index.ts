export { h } from './h'
export { renderSlots } from './helpers/renderSlots'
export { createTextVNode, createVNode as createElementVNode } from './vnode'
export { getCurrentInstance, registerCompiler } from './component'
export { provide, inject } from './apiInject'
export { createRenderer } from './render'
export { nextTick } from './scheduler'
export { toDisplayString } from '../shared'
