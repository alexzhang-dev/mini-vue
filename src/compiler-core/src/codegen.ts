import { NodeType } from './ast'
import {
  CREATE_ELEMENT_VNODE,
  HelperNameMapping,
  TO_DISPLAY_STRING,
} from './runtimeHelpers'

export function codegen(ast) {
  const context = createCodegenContext()
  const { push } = context

  if (ast.helpers.length) {
    genFunctionPreamble(ast, context)
  }

  const funcName = 'render'
  push(`export `)
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')
  push(`function ${funcName}(${signature}) { `)
  push(`return `)
  genNode(ast.codegenNode, context)
  push(` }`)
  return context.code
}

function genFunctionPreamble(ast, context) {
  const VueBinding = 'Vue'
  const { push, newLine } = context
  const aliasHelper = s => `${HelperNameMapping[s]}: _${HelperNameMapping[s]}`
  push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinding}`)
  newLine()
}

function genNode(node, context) {
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context)
      break
    case NodeType.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeType.ELEMENT:
      genElement(node, context)
      break
  }
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}('${tag}')`)
}

function genExpression(node, context) {
  const { push } = context
  push(`${node.content}`)
}

function genInterpolation(node, context) {
  const { push } = context
  push(`${context.helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(`)`)
}

function genText(node, context) {
  const { push } = context
  push(`'${node.content}'`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source: string) {
      context.code += source
    },
    newLine() {
      context.code += '\n'
    },
    helper(name) {
      return `_${HelperNameMapping[name]}`
    },
  }
  return context
}
