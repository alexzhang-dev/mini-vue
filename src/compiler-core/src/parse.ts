import { NodeType } from './ast'

const enum TagType {
  START,
  END,
}

export function baseParse(content: string) {
  const context = createContext(content)
  return createRoot(parseChildren(context))
}

function createContext(content: string) {
  return {
    source: content,
  }
}

function createRoot(children) {
  return {
    children,
  }
}

function parseChildren(context: { source: string }): any {
  const nodes: any = []
  let node
  const s = context.source
  if (s.startsWith('{{')) {
    node = parseInterpolation(context)
  } else if (s.startsWith('<') && /[a-z]/i.test(s[1])) {
    node = parseElement(context)
  }
  nodes.push(node)
  return [node]
}

function parseInterpolation(context: { source: string }) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'
  // 将字符串截取为 message}}
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  )
  // 让后将字符串前面的 {{ 舍弃掉，我们将其称之为【推进】
  advanceBy(context, openDelimiter.length)
  // 获取到 {{}} 中间值的长度
  const rawContentLength = closeIndex - openDelimiter.length
  // 并将中间这个值获取出来
  const rawContent = context.source.slice(0, rawContentLength)
  const content = rawContent.trim()
  // 继续【推进】
  advanceBy(context, rawContentLength + closeDelimiter.length)
  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content: content,
    },
  }
}

function advanceBy(context, length: number) {
  context.source = context.source.slice(length)
}

function parseElement(context: { source: string }): any {
  const element = parseTag(context, TagType.START)
  parseTag(context, TagType.END)
  return element
}

function parseTag(context: { source: string }, type: TagType) {
  // i 忽略大小写, ([a-z]*) 作为一个分组
  const match = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match![1]
  advanceBy(context, match![0].length + 1)
  if (type === TagType.END) return
  return {
    type: NodeType.ELEMENT,
    tag,
  }
}
