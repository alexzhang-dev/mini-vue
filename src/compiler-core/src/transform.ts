import { NodeType } from './ast'
import { HelperNameMapping, TO_DISPLAY_STRING } from './runtimeHelpers'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)
  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
  root.codegenNode = root.children[0]
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || {},
    helpers: new Map(),
    helper(name: string) {
      context.helpers.set(name, 1)
    },
  }
  return context
}

function traverseNode(node, context) {
  const { nodeTransforms } = context
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node, context)
  }
  switch (node.type) {
    case NodeType.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeType.ROOT:
    case NodeType.ELEMENT:
      traverseChildren(node, context)
      break
    default:
      break
  }
}

function traverseChildren(node, context) {
  const children = node.children
  for (let i = 0; i < children.length; i++) {
    traverseNode(children[i], context)
  }
}
