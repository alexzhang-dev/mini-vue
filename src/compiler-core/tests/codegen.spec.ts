import { codegen } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { transformExpression } from '../src/transfroms/transformExpression'

describe('codegen', () => {
  test('text', () => {
    const template = 'hi'
    const ast = baseParse(template)
    transform(ast)
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
  test('interpolation', () => {
    const template = '{{message}}'
    const ast = baseParse(template)
    transform(ast, {
      nodeTransforms: [transformExpression],
    })
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
})
