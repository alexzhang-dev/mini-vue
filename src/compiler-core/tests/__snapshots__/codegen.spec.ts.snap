// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
export function render(_ctx, _cache) { return _toDisplayString(_ctx.message) }"
`;

exports[`codegen simple element 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
export function render(_ctx, _cache) { return  }"
`;

exports[`codegen text 1`] = `"export function render(_ctx, _cache) { return 'hi' }"`;
